### PR TITLE
Throw exception when calling FunctionRegistry#getUdfFactory(name) if function doesn't exist

### PR DESCRIPTION
**Pull request by dguy**

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java [hunk 1]
```diff
@@ -191,8 +191,13 @@ protected Expression visitSubscriptExpression(
   protected Expression visitFunctionCall(final FunctionCall node,
                                          final ExpressionTypeContext expressionTypeContext) {
 
-    final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName().getSuffix());
-    if (udfFactory != null) {
+    if (functionRegistry.isAggregate(node.getName().getSuffix())) {
+      KsqlAggregateFunction ksqlAggregateFunction =
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java]: nit: final

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 2]
```diff
@@ -75,7 +75,11 @@ private void init() {
   }
 
   public UdfFactory getUdfFactory(final String functionName) {
-    return ksqlFunctionMap.get(functionName.toUpperCase());
+    final UdfFactory udfFactory = ksqlFunctionMap.get(functionName.toUpperCase());
+    if (udfFactory == null) {
+      throw new KsqlException("Can't find any functions with the name " + functionName);
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: nit: Function name in single quotes?

File: ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java [hunk 3]
```diff
@@ -94,7 +94,12 @@ public void shouldCreateCopyOfRegistry() {
 
     assertThat(copy.getUdfFactory("func").getFunction(Collections.emptyList()), equalTo(func));
     assertThat(copy.getUdfFactory("func2").getFunction(Collections.emptyList()), equalTo(func2));
-    assertThat(functionRegistry.getUdfFactory("func2"), nullValue());
+    try {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java]: This duplicates `shouldThrowExceptionIfNoFunctionsWithNameExist` testing, right?

File: ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java [hunk 4]
```diff
@@ -176,4 +181,9 @@ public void shouldAddFunctionWithSameNameClassButDifferentArguments() {
     assertThat(functionRegistry.getUdfFactory("func")
         .getFunction(Collections.emptyList()), equalTo(func));
   }
+
+  @Test(expected = KsqlException.class)
+  public void shouldThrowExceptionIfNoFunctionsWithNameExist() {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java]: Maybe use ExpectedException to test the error message contains the function name

File: ksql-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java [hunk 5]
```diff
@@ -129,12 +131,16 @@ public void shouldNotLoadUdfsInJarDirectoryIfLoadCustomerUdfsFalse() {
     pluginLoader.load();
     // udf in ksql-engine
     final UdfFactory function = metaStore.getUdfFactory("substring");
+    assertThat(function, not(nullValue()));
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java]: Can't be null though, right? So isn't this assert pointless?

File: ksql-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java [hunk 6]
```diff
@@ -129,12 +131,16 @@ public void shouldNotLoadUdfsInJarDirectoryIfLoadCustomerUdfsFalse() {
...
+
     // udf in udf-example.jar
-    final UdfFactory toString = metaStore.getUdfFactory("tostring");
+    try {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/function/UdfLoaderTest.java]: Feels like it should be in a separate test.

File: ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java (hunk 3 above)
Comment (dguy) [reply ksql-engine/src/test/java/io/confluent/ksql/function/InternalFunctionRegistryTest.java]: It does, but proves that the new function doesn't get added to the registry it was copied from. Worth keeping both

File: ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java [hunk 7]
```diff
@@ -191,23 +191,21 @@ protected Expression visitSubscriptExpression(
...
                                          final ExpressionTypeContext expressionTypeContext) {
 
-    final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName().getSuffix());
-    if (udfFactory != null) {
-      List<Schema.Type> argTypes = new ArrayList<>();
+    if (functionRegistry.isAggregate(node.getName().getSuffix())) {
+      final KsqlAggregateFunction ksqlAggregateFunction =
+          functionRegistry.getAggregate(
+              node.getName().getSuffix(), getExpressionSchema(node.getArguments().get(0)));
+      expressionTypeContext.setSchema(ksqlAggregateFunction.getReturnType());
```
Comment (hjafarpour) [inline ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java]: Wouldn't we have the same problem when we have UDAFs?

Comment (dguy) [reply ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java]: We will, but at the moment the call `functionRegistry.isAggregate(node.getName().getSuffix())` will return false, so there will be no null pointer. Once we add the code to handle UDAFs we will need to update this also.

**Pull request merged**